flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1763,6 +1765,66 @@ public long calculateThroughput() {
         }
     }
 
+    /**
+     * Tests mailbox metrics latency and queue size and verifies that (1) latency measurement is
+     * executed initially once and at least once triggered by timer, (2) latency mean value is
+     * greater than zero and (3) mailbox size is greater than zero for some time and eventually
+     * equals to zero.
+     *
+     * @throws Exception on {@link MockEnvironmentBuilder#build()} failure.
+     */
+    @Test
+    public void testMailboxMetrics() throws Exception {
```

Comment:
Are you sure this test is going to be stable? It looks like 20ms freeze of the metrics thread would cause it to fail. 

Can you take a look at the buffer debloating tests and `StreamTask#debloat` method in particular? I think using `StreamTaskMailboxTestHarnessBuilder` you would be able to implement a precise and reliable test. 

1. Unlike `StreamTask#debloat` method, yours `StreamTask#meassureMailboxLatency` would look like:
```
StreamTask#meassureMailboxLatency() {
                    long startTime = SystemClock.getInstance().relativeTimeMillis();
                    mainMailboxExecutor.execute(
                            () -> {
                                long endTime = SystemClock.getInstance().relativeTimeMillis();
                                long latency = endTime - startTime;
                                environment
                                        .getMetricGroup()
                                        .getIOMetricGroup()
                                        .getMailboxLatency()
                                        .update(latency);
                                scheduleMailboxMetrics();
                            },
                            "Measure mailbox latency metric");
                });
}
```
2. You could enqueue 10 mails.
3. Setup to sleep X ms (`20ms`?)  per record.
4. Call `StreamTask#meassureMailboxLatency`
5. Assert `mailboxQueueSize == 11`.
6. Call `testHarness.processAll()`
7. Assert `maxLatency >= X ms`.

Reply:
1. The test is stable in the way that any additional delay will not cause the test to fail. Note that the assertions ask for a minimum delay and minimum number of latency measurements (>= 2) and `Thread.sleep(iteration == 2 ? 10 : 1);` is only to not let the test sleep/run for too long, e.g., total execution time is currently ~34ms. (The test would be work the same way if there was only `Thread.sleep(10) or `Thread.sleep(20)`.
2. I don't see the point of testing with testHarness.processMail because 1) it does not call the StreamTask#invoke method and, hence, won't perform the initial call of the scheduleMailboxMetrics and 2) I don't see/understand how the testHarness uses/implements timer-triggered execution of actions which is relevant part of the code to be tested.

Reply:
With this simple modification:
```
    private void scheduleMailboxMetrics() {
        systemTimerService.registerTimer(
                systemTimerService.getCurrentProcessingTime() + mailboxMetricsInterval,
                timestamp -> {
                    long startTime = SystemClock.getInstance().relativeTimeMillis();
                    Thread.sleep(2000);  // <<<<< added sleep
                    mainMailboxExecutor.execute(...);
    }
```
the test is failing so you can see it won't be stable in case of JVM hiccups.

> it does not call the StreamTask#invoke method

Yes, that's the downside. If needed we can have detailed correctness test (a) implemented with the mailbox test harness, where we can do proper assertions easily complemented with a basic test (b) that performs some trivial assertion that `scheduleMailboxMetrics()` has been called in the `StreamTask#invoke`. (a) would be as I described above. (b) could be a simplified version of this test, that just do a while true loop as long as reported latency is > 0ms.